transport/awssqs/consumer.go

```diff
@@ -0,0 +1,227 @@
+}
+
+// ConsumerAfter functions are executed on the consumer reply after the
+// endpoint is invoked, but before anything is published to the reply.
+func ConsumerAfter(after ...ConsumerResponseFunc) ConsumerOption {
+	return func(c *Consumer) { c.after = append(c.after, after...) }
+}
+
+// ConsumerErrorEncoder is used to encode errors to the consumer reply
+// whenever they're encountered in the processing of a request. Clients can
+// use this to provide custom error formatting. By default,
+// errors will be published with the DefaultErrorEncoder.
+func ConsumerErrorEncoder(ee ErrorEncoder) ConsumerOption {
+	return func(c *Consumer) { c.errorEncoder = ee }
+}
+
+// ConsumerWantReplyFunc overrides the default value for the consumer's
+// wantRep field.
+func ConsumerWantReplyFunc(replyFunc WantReplyFunc) ConsumerOption {
+	return func(c *Consumer) { c.wantRep = replyFunc }
+}
+
+// ConsumerErrorHandler is used to handle non-terminal errors. By default, non-terminal errors
+// are ignored. This is intended as a diagnostic measure. Finer-grained control
+// of error handling, including logging in more detail, should be performed in a
+// custom ConsumerErrorEncoder which has access to the context.
+func ConsumerErrorHandler(errorHandler transport.ErrorHandler) ConsumerOption {
+	return func(c *Consumer) { c.errorHandler = errorHandler }
+}
+
+// ConsumerFinalizer is executed once all the received SQS messages are done being processed.
+// By default, no finalizer is registered.
+func ConsumerFinalizer(f ...ConsumerFinalizerFunc) ConsumerOption {
+	return func(c *Consumer) { c.finalizer = f }
+}
+
+// ConsumerDeleteMessageBefore returns a ConsumerOption that appends a function
+// that delete the message from queue to the list of consumer's before functions.
+func ConsumerDeleteMessageBefore() ConsumerOption {
+	return func(c *Consumer) {
+		deleteBefore := func(ctx context.Context, cancel context.CancelFunc, msg *sqs.Message) context.Context {
+			if err := deleteMessage(ctx, c.sqsClient, c.queueURL, msg); err != nil {
+				c.errorHandler.Handle(ctx, err)
+				c.errorEncoder(ctx, err, msg, c.sqsClient)
+				cancel()
+			}
+			return ctx
+		}
+		c.before = append(c.before, deleteBefore)
+	}
+}
+
+// ConsumerDeleteMessageAfter returns a ConsumerOption that appends a function
+// that delete a message from queue to the list of consumer's after functions.
+func ConsumerDeleteMessageAfter() ConsumerOption {
+	return func(c *Consumer) {
+		deleteAfter := func(ctx context.Context, cancel context.CancelFunc, msg *sqs.Message, _ *sqs.SendMessageInput) context.Context {
+			if err := deleteMessage(ctx, c.sqsClient, c.queueURL, msg); err != nil {
+				c.errorHandler.Handle(ctx, err)
+				c.errorEncoder(ctx, err, msg, c.sqsClient)
+				cancel()
+			}
+			return ctx
+		}
+		c.after = append(c.after, deleteAfter)
+	}
+}
+
+// ServeMessage serves an SQS message.
+func (c Consumer) ServeMessage(ctx context.Context, msg *sqs.Message) error {
```

Comment:
Does this implement an interface expected of something in the AWS SQS package? How should callers use it?

Reply:
No not really.
It should be used as depicted by @xyluet in https://github.com/go-kit/kit/pull/1018#issuecomment-707457997

Reply:
I think there is some conceptual disconnect here. 

A Go kit service is comprised of 1 or more endpoints, and each endpoint is exposed via 1 or more transports. By convention every Go kit transport package provides a type that exposes a single endpoint with unique DecodeRequest and EncodeResponse functions. But that type should compose into a larger "unit" which represents an entire service. 

For example, transport/http.Server type isn't an sttdlib http.Server itself but an http.Handler, and you're supposed to mount multiple transport/http.Server handlers in a single mux to represent your service. Or, transport/nats.Subscriber isn't it's own client and consumer of the NATS topic, instead it implements nats.MsgHandler so that it can be composed by the caller into a larger consumer that manages the whole service.

Concretely: users of this package shouldn't have to run ServeMessage loops for every transport/awssqs.Consumer they create (i.e. every endpoint in their service). They should create one SQS client/consumer/whatever with 1 or more transport/awssqs.Consumer types, each of which is fed messages by the "outer" component appropriately. I don't know the architecture of the SQS client lib so I don't know if this is natively supported or would have to be provided in this package too.

Does this make sense?

Reply:
It doesn't implement anything. I just borrow from amqp transport.
https://github.com/go-kit/kit/blob/6c170213364cbdae29ce60c4a505ec24c886172f/transport/amqp/subscriber.go#L98

SQS SDK provides almost the same way as RabbitMQ when receiving messages. RabbitMQ is using channel, and we need to for loop the channel and pass a `Delivery`. SQS has the same way, but we need to call `ReceiveMessage` in for loop and getting array of messages from receive message output.

Reply:
any update on how you feel about this, knowing that this pattern is already used in amqp transport @peterbourgon ?

Reply:
Oh, man, I typed a big comment in reply but I guess it didn't save? :( Will re-type in a bit here, sorry!

Reply:
An important difference between AMQP's ServeDelivery and your ServeMessage is that ServeDelivery doesn't "run" — it returns a function that takes a received amqp.Delivery, runs through the request and response flow, and then emits a response to the provided amqp.Channel. Users need to feed the returned function deliveries, but they can do that however they like, they don't necessarily need to manage a goroutine per Subscriber to do so.

So ServeMessage could work that way, maybe. Or maybe there's a better fit. But the package shouldn't require callers to `go consumer.Anything()` to work properly.

Reply:
So ServeMessage should return function like AMQP?
```go 
ServeMessage(<maybe SQS client and/or cancellation context>) func(msg *sqs.Message) 
```

Reply:
If that works with the SQS SDK, then sure. Maybe another approach is better. I can't judge.

Reply:
I don't really see how this solves the problem, you'd still end up doing something like

``` go
subscriber := sqstransport.NewSubscriber(endpoint, dec, enc, opts...)

for {
  out, _ := sqsClient.ReceiveMessage(ctx, input) //handle error
  for _, msg := range out.Messages {
    consumer.ServeMessage(ctx)(msg)
  }
}
```

or

``` go
subscriber := sqstransport.NewSubscriber(endpoint, dec, enc, opts...)
serveFunc := consumer.ServeMessage(ctx)

for {
  out, _ := sqsClient.ReceiveMessage(ctx, input) //handle error
  for _, msg := range out.Messages {
    serveFunc(msg)
  }
}
```

Is there something I didn't catch ?

The big difference with amqp is that, as far as I know, the SQS SDK does not provide a function that you can call to setup a function that will serve arriving messages automatically.

You need to fetch the messages by your self and handle them on your own. So we have two possibilities :
* We include fetching the messages in go-kit but this would mean including complex code to allow visibility timeout extension and other SQS features that are necessary for some users.
* We let the end user fetch the messages and calling the function that serves the messages.

In both cases the user will need to call `c.ServeMessage` or `c.Consume`, I personally see no other way.

How about you @xyluet ?

Reply:
Just like AMQP, they could just pass `Channel` and call `Consume` inside. 

I prefer let the end user fetch the messages and calling the function that serves a message. We can still use context with closure for other things, e.g. for acking before or after endpoint, etc.

Reply:
I just changed the ServeMessage method signature to return a function. Is this what both of you had in mind ?

Reply:
Any update guys ?

Reply:
Hello !

What are your thoughts about the last changes I committed https://github.com/go-kit/kit/pull/1018/commits/b673cbefc4406a0cbfc7f29de081fc62042316f3 ?

Reply:
@xyluet @peterbourgon Any update on this ?

Reply:
> I think there is some conceptual disconnect here.
> 
> A Go kit service is comprised of 1 or more endpoints, and each endpoint is exposed via 1 or more transports. By convention every Go kit transport package provides a type that exposes a single endpoint with unique DecodeRequest and EncodeResponse functions. But that type should compose into a larger "unit" which represents an entire service.
> 
> For example, transport/http.Server type isn't an sttdlib http.Server itself but an http.Handler, and you're supposed to mount multiple transport/http.Server handlers in a single mux to represent your service. Or, transport/nats.Subscriber isn't it's own client and consumer of the NATS topic, instead it implements nats.MsgHandler so that it can be composed by the caller into a larger consumer that receives multiple message types and dispatches them to the appropriate MsgHandler.
> 
> Concretely: users of this package shouldn't have to run ServeMessage loops for every transport/awssqs.Consumer they create (i.e. every endpoint in their service). They should create one SQS client/consumer/whatever with 1 or more transport/awssqs.Consumer types, each of which is fed messages by the "outer" component appropriately. I don't know the architecture of the SQS client lib so I don't know if this is natively supported or would have to be provided in this package too.
> 
> Does this make sense?

SQS has this concept of long polling and short polling. Long polling would be a long lived connection that lives for a maximum of 20 seconds, allowing the poller to return a response only when there is a message available (cheaper due to less network calls). With short polling, a response is returned immediately but may have an empty response(may require multiple requests to check for a message on the queue).

This is the receive message api endpoint from sqs: 

```ReceiveMessage(input *ReceiveMessageInput) (*ReceiveMessageOutput, error)```

```
// MsgHandler is a callback function that processes messages delivered to
// asynchronous subscribers.
type MsgHandler func(msg *Msg)
```